### DISCUSSION TITLE
[JS]이벤트 버블링,이벤트 캡처링에 대해 설명하세요.

### DISCUSSION THREAD

**Discussion by bona373737** [wanted-9th-3team/frontend-study#14] [프론트 기술 질문 공간]
이벤트 버블링, 이벤트 캡처링

**Comment by javascriptbiggosoo** [comment]
특정 element(target)에서 클릭 이벤트가 일어나는 경우, 이벤트는 다음과 같은 순서로 전파됩니다.
1. document root에서 클릭 이벤트를 생성합니다
2. 이후 target의 부모 요소들을 순차적으로 거쳐갑니다 이를 캡처링 페이즈라고 합니다
3. 클릭 이벤트가 target에 도달합니다. 이를 타겟 페이즈라고 합니다.
4. 역순으로 부모 요소들을 타고 올라갑니다. 이를 버블링 페이즈라고 합니다.


이게 이벤트의 전파경로이며 기본적으로 부모요소들은 캡처링, 버블링 각 페이즈마다 한 번씩 이벤트가 거쳐갑니다.
때문에 이벤트를 다룰 때엔 3가지 페이즈 중 적절한 때에 이벤트를 청취하도록 조정하는 것이 중요했는데요.
브라우저가 서로 웹표준을 지키지 않은 때에는 순서가 뒤바뀌는 경우가 발생할 수 있어 이것을 염두에 두는게 중요했습니다.
지금은 대부분의 브라우저에서 버블링 페이즈에 이벤트를 수신하는 것이 디폴트로 설정되어있어 특별한 경우가 아니라면 신경쓸 일이 없습니다

**Comment by michoball** [comment]
특정 element에 이벤트를 등록하면 사용자가 해당 이벤트를 발생시킬때 
최상위 root dom tree에서 부터 해당 이벤트가 발생한 element를 찾아 아래로 타고 들어가는 과정을 캡처링이라고 합니다.

해당 element를 찾으면 element에서 이벤트가 발생하고 root element로 이벤트가 전파되는 것을 버블링이라고 합니다. 이벤트의 전파를 막고 싶으면 stopPropagation 메서드를 호출하면 됩니다.

이 이벤트가 버블링 되는 현상을 이용하여 여러 하위 element 에 이벤트를 등록하지 않고 상위 element 에 이벤트를 등록하여 사용하는 방식을 이벤트 위임이라고 합니다.

**Comment by skyhanull** [comment]
이벤트 버블링
- 이벤트 버블링은 특정 화면 요소에서 이벤트가 발생했을 때 해당 이벤트가 더 상위의 화면 요소들로 전달되어 가는 특성을 의미합니다.

이벤트 캡쳐링
- 이벤트 캡쳐는 이벤트 버블링과 반대 방향으로 진행되는 이벤트 전파 방식입니다.

버블링과 캡쳐링을 막을 수 있는 메소드로는 Event.stopPropagation()를 사용하여 막을 수 있습니다

**Comment by WooGie911** [comment]
이벤트 버블링과 캡처링은 돔 트리상에 존재하는 모든 돔 요소 노드에서 발생한 이벤트가 돔트리를 통해 전파되는 이벤트 전파 방식중에 하나입니다. 
전파되는 방향에 따라 캡처링, 타깃, 버블링 3단계로 구분됩니다. 
특정 타깃에서 이벤트가 발생하게 되면, 최상위 노드(root)에서 부터 캡처링이 일어나서 타깃 요소까지 이벤트가 전해져 내려옵니다. 
이벤트는 타깃요소에서 부터 다시 최상위 노드(root)까지 버블링이 일어나서 이벤트가 전달됩니다. 
우리는 일반적으로 버블링 현상을 캐치하여 사용합니다. 
이벤트를 상위 노드에 위임하여 하위 노드의 이벤트를 조작하는 것이 그 예시입니다. 
이벤트 리스너의 세번째 인자값 값인 capture 옵션(boolean)을 true로 선언하면 캡쳐링 단계의 이벤트도 캐치할 수 있습니다.
